src/toil/cwl/cwltoil.py

```diff
@@ -757,15 +797,58 @@ def __init__(self, tool, cwljob, runtime_context, step_inputs=None, conditional=
             # an identifier
             pass
         self.runtime_context = runtime_context
-        self.step_inputs = step_inputs or self.cwltool.tool["inputs"]
+        self.step_inputs = self.cwltool.tool["inputs"]
         self.workdir = runtime_context.workdir
 
-    def run(self, file_store):
-        cwljob = resolve_indirect(self.cwljob)
+    def required_env_vars(self, cwljob: Any) -> Iterator[Tuple[str, str]]:
+        """An iterator that yields environment variables specified with the EnvVarRequirement keyword."""
+        if isinstance(cwljob, dict):
+            if cwljob.get("class") == 'EnvVarRequirement':
+                for t in cwljob.get("envDef", {}):
+                    yield t["envName"], self.builder.do_eval(t["envValue"])
+            for k, v in cwljob.items():
+                for env_name, env_value in self.required_env_vars(v):
+                    yield env_name, env_value
+        if isinstance(cwljob, list):
+            for env_var in cwljob:
+                for env_name, env_value in self.required_env_vars(env_var):
+                    yield env_name, env_value
+
+    def populate_env_vars(self, cwljob: dict) -> dict:
+        """
+        Prepares environment variables necessary at runtime of the job.
+
+        Env vars specified in the CWL "requirements" section should already be loaded in self.cwltool.requirements,
+        however those specified with "EnvVarRequirement" take precedence and are only populated here.  Therefore,
+        this not only returns a dictionary with all evaluated "EnvVarRequirement" env vars, but checks
+        self.cwltool.requirements for any env vars with the same name and replaces their value with that found in the
+        "EnvVarRequirement" env var if it exists.
+        """
+        self.builder.job = cwljob
+        required_env_vars = {}
+        # iterate over EnvVarRequirement env vars, if any
+        for k, v in self.required_env_vars(cwljob):
+            required_env_vars[k] = v  # will tell cwltool which env vars to take from the environment
+            os.environ[k] = v  # needs to actually be populated in the environment as well or they're not used
+
+        # EnvVarRequirement env vars take priority over those specified with "requirements"
+        # so cwltool.requirements need to be overwritten if an env var with the same name is found
+        for req in self.cwltool.requirements:
+            for env_def in req.get('envDef', {}):
+                env_name = env_def.get('envName', '')
+                if env_name in required_env_vars:
+                    env_def['envValue'] = required_env_vars[env_name]
```

Comment:
You're sure that modifying the "requirements" env vars won't affect any other jobs? Jobs have both "requirements" and "EnvVarRequirements" just as part of the job, but if they conflict (because the author weirdly tried to use them both), "EnvVarRequirements" win?

Reply:
Test 228 is the one where the author weirdly tried to use both: `Test conflicting requirements in input document via EnvVarRequirement and expression`

So I'm overriding based on that test.  I'm relying a bit on the conformance tests to help me understand what needs to work in what way, but this seems to be how cwltool overrides.  Maybe @mr-c can comment?